allennlp/tests/training/event_based_trainer_test.py

```diff
@@ -0,0 +1,131 @@
+# pylint: disable=invalid-name,too-many-public-methods,protected-access
+from typing import Iterable
+
+import torch
+
+from allennlp.common.params import Params
+from allennlp.common.testing import AllenNlpTestCase
+from allennlp.common.tqdm import Tqdm
+from allennlp.data import Vocabulary
+from allennlp.data.dataset_readers import SequenceTaggingDatasetReader
+from allennlp.data.iterators import BasicIterator
+from allennlp.data.iterators.data_iterator import TensorDict
+from allennlp.models.simple_tagger import SimpleTagger
+from allennlp.training import Trainer
+from allennlp.training.event_based_trainer.callback import Callback
+from allennlp.training.event_based_trainer.event_based_trainer import EventBasedTrainer
+from allennlp.training.event_based_trainer.events import Events
+
+
+class TrainOneBatch(Callback):
+    """
+    Trains one batch by running
+        optimizer.zero_grad -> model.forward -> loss.backward -> optimizer.step
+    """
+    def __init__(self, optimizer):
+        self.optimizer = optimizer
+
+    def call(self, trainer: EventBasedTrainer) -> None:
+        self.optimizer.zero_grad()
+
+        trainer.output = trainer.model(**trainer.batch)
+        loss = trainer.output["loss"]
+        loss.backward()
+        trainer.tqdm.set_description(f"loss: {loss}")
+
+        self.optimizer.step()
+
+
+class Evaluate(Callback):
+    def __init__(self,
+                 epoch_generator: Iterable[TensorDict],
```

Comment:
This one, also; getting the data into here `from_params` is probably going to tricky, but quite easy if it just lives in the trainer state.

Reply:
same as above, I don't think you'd ever want to instantiate `EventBasedTrainer` from_params (maybe I should call it `EventBasedTrainerBase`? yuck), I wanted to build the minimal base class that you could easily build any variety of trainers on top of.

Reply:
No, I think you do want to instantiate this class `FromParams`, because all of the fun stuff is just happening in the callbacks, anyway.  You're right, though, that these some of these things do need to have some particularly special magic to make their way to the right callbacks (particularly the optimizers, which need the model parameters).